cvmfs/swissknife_check.cc

```diff
@@ -566,15 +583,22 @@ bool CommandCheck::Find(const catalog::Catalog *catalog,
         // are all data chunks in the data store?
         if (check_chunks_ && !entries[i].IsExternalFile()) {
           const shash::Any &chunk_hash = this_chunk.content_hash();
-          const string chunk_path = "data/" + chunk_hash.MakePath();
-          if (!Exists(chunk_path)) {
-            LogCvmfs(kLogCvmfs, kLogStderr, "partial data chunk %s (%s -> "
-                                            "offset: %d | size: %d) missing",
-                     this_chunk.content_hash().ToStringWithSuffix().c_str(),
-                     full_path.c_str(),
-                     this_chunk.offset(),
-                     this_chunk.size());
-            retval = false;
+          // for performance reasons, only perform the check once
+          // and skip if the hash has been checked before
+          if (duplicates_map_.Contains(chunk_hash)) {
```

Comment:
```suggestion
          if (!duplicates_map_.Contains(chunk_hash)) {
            duplicates_map_.Insert(chunk_hash, 1);
```
and then no `else` branch.

NB: This method has grown uncomfortably complicated. Perhaps let's add a TODO to split it into smaller methods.

Reply:
Indeed that was the wrong way around, thanks!